File: Ruby/ruby_solutions_11-20/Euler_14.rb
#!/usr/bin/env ruby


require 'timeout'


def collatz(n)
  count = 0
  while n != 1
    if n.modulo(2) != 0
      n *= 3
      n += 1
      count += 1
    end 
    if n.modulo(2) == 0
      n = n >> 1
      count += 1
    end
  end
  count
end

def Euler_14
  answer = 0
  high = 0
  for i in (3..1000000).step(2)
    canidate = collatz(i)
    if canidate > high 
      answer = i 
      high = canidate
    end
  end
  puts "Answer: #{answer}"
end

if __FILE__ == $0
  start = Time.now()
  Euler_14()
  finish = Time.now() - start
  puts "Time: #{finish}"
end

File: C/c_solutions_11-20/Euler_14.c
#include "euler_util.h"


int main(int argc, char *argv[])
{
    float start = timeit();

    long long high_num = 0;

    for (long long canidate=3, high=0, count=0; canidate < 1000000; canidate += 2, count=0) {
        for (long long collatz=canidate; collatz != 1; count++) 
            collatz = collatz % 2 == 0 ? collatz / 2 : (collatz * 3) + 1; 
        if (count > high) {
            high = count;
            high_num = canidate;
        }
    }

    float stop = timeit();

    printf ("Answer: %lld\n", high_num);
    printf ("Time: %f\n", stop - start);

    return 0;
}
        
        
